adding KPI's DG, OEF and OEM

Piranias opened the pull request:
Fix #603

**Changes proposed in this pull request**:
- in E3: added functions: 

```
* add_degree_of_autonomy() and equation_degree_of_autonomy()
* add_onsite_energy_fraction() and equation_onsite_energy_fraction()
* add_add_add_onsite_energy_matching() and equation_onsite_energy_matching()
```

- write tests for each function in Test_E3
- add to changelog
- added constants to constants_json_strings.py
- added the three KPI functions to E0


The following steps were realized, as well (if applies):
- [x] Use in-line comments to explain your code
- [x] Write docstrings to your code ([example docstring](https://multi-vector-simulator.readthedocs.io/en/latest/Developing.html#format-of-docstrings))
- [x] For new functionalities: Explain in readthedocs
- [x] Write test(s) for your new patch of code (pytests, assertion debug messages)
- [x] Update the CHANGELOG.md
- [x] Apply black (`black . --exclude docs/`)
- [x] Check if benchmark tests pass locally (`EXECUTE_TESTS_ON=master pytest`)


*Please mark above checkboxes as following:*
- [ ] Open
- [x] Done

:x: Check not applicable to this PR

<sub>*For more information on how to contribute check the [CONTRIBUTING.md](https://github.com/rl-institut/multi-vector-simulator/blob/dev/CONTRIBUTING.md).*<sub>

Piranias commented:
unfortunately I changed the branch too late, apparently I transferred changes from PR #609 (read the docs) to this PR without noticing.
This should be solved when PR #609 is merged..

SabineHaas commented on src/multi_vector_simulator/E3_indicator_calculation.py:
```diff
@@ -417,7 +425,75 @@ def equation_renewable_share(total_res, total_non_res):
+
+def equation_degree_of_autonomy(total_generation, total_demand):
+    """
+    calculates the degree of autonomy.
```
```suggestion
    Calculates the degree of autonomy (DA).
```

SabineHaas commented on src/multi_vector_simulator/E3_indicator_calculation.py:
```diff
@@ -488,6 +564,7 @@ def equation_degree_of_sector_coupling(
     Returns
     -------
+    float
     Degree of sector coupling based on conversion flows and energy demands in electricity equivalent.
```
```suggestion
        Degree of sector coupling based on conversion flows and energy demands in electricity equivalent.
```

SabineHaas commented on src/multi_vector_simulator/E3_indicator_calculation.py:
```diff
@@ -502,11 +579,177 @@ def equation_degree_of_sector_coupling(
-def equation_onsite_energy_fraction():
+def add_onsite_energy_fraction(dict_values):
+    """
+    Determines onsite energy fraction /self consumption and adds KPI to dict_values
```
```suggestion
    Determines onsite energy fraction (OEF), i.e. self-consumption, and adds KPI to dict_values
```

SabineHaas commented on src/multi_vector_simulator/E3_indicator_calculation.py:
```diff
@@ -502,11 +579,177 @@ def equation_degree_of_sector_coupling(
+    dict_values: dict
+        dict with all project information and results
+        after applying total_renewable_and_non_renewable_energy_origin
+    Returns
```
```suggestion
    
    Returns
```

SabineHaas commented on src/multi_vector_simulator/E3_indicator_calculation.py:
```diff
@@ -502,11 +579,177 @@ def equation_degree_of_sector_coupling(
+
+
+def equation_onsite_energy_fraction(total_generation, total_feedin):
+    """Calculates onsite energy fraction / self-consumption.
```
```suggestion
    """
    Calculates onsite energy fraction (OEF), i.e. self-consumption.
```

SabineHaas commented on src/multi_vector_simulator/E3_indicator_calculation.py:
```diff
@@ -502,11 +579,177 @@ def equation_degree_of_sector_coupling(
+        # load total flow into the dso sink
+        dso_sink = scalars_matrix.loc[
+            scalars_matrix["label"]
+            == dict_values[ENERGY_PROVIDERS][  # todo: IS this referece always correct??
```
@smartie2076  could you check this please?

smartie2076 replied:
Is `"label"` here the same as `LABEL` or a different one?

smartie2076 replied:
Yeah always correct.

SabineHaas commented on src/multi_vector_simulator/E3_indicator_calculation.py:
```diff
@@ -502,11 +579,177 @@ def equation_degree_of_sector_coupling(
+        total_generation, total_feedin
+    )
+
+    # save total feedin into KPI_SCALARS_DICT
```
```suggestion
    # save KPI  onsite energy fraction into KPI_SCALARS_DICT
```

SabineHaas commented on src/multi_vector_simulator/E3_indicator_calculation.py:
```diff
@@ -417,7 +425,75 @@ def equation_renewable_share(total_res, total_non_res):
+    dict_values[KPI][KPI_SCALARS_DICT].update({DEGREE_OF_AUTONOMY: degree_of_autonomy})
+
+    logging.debug(f"Calculated the {DEGREE_OF_AUTONOMY}: {round(degree_of_autonomy)}")
+    logging.info(f"Calculated the {DEGREE_OF_AUTONOMY}.")
```
```suggestion
    logging.info(f"Calculated the {DEGREE_OF_AUTONOMY} of the LES.")
```

SabineHaas commented on src/multi_vector_simulator/E3_indicator_calculation.py:
```diff
@@ -502,11 +579,177 @@ def equation_degree_of_sector_coupling(
+    logging.debug(
+        f"Calculated the {ONSITE_ENERGY_FRACTION}: {round(onsite_energy_fraction)}"
+    )
+    logging.info(f"Calculated the {ONSITE_ENERGY_FRACTION}.")
```
```suggestion
    logging.info(f"Calculated the {ONSITE_ENERGY_FRACTION} of the LES.")
```

SabineHaas commented on src/multi_vector_simulator/E3_indicator_calculation.py:
```diff
@@ -502,11 +579,177 @@ def equation_degree_of_sector_coupling(
+    logging.debug(
+        f"Calculated the {ONSITE_ENERGY_MATCHING}: {round(onsite_energy_matching)}"
+    )
+    logging.info(f"Calculated the {ONSITE_ENERGY_MATCHING}.")
```
```suggestion
    logging.info(f"Calculated the {ONSITE_ENERGY_MATCHING} of the LES.")
```

SabineHaas commented on src/multi_vector_simulator/E3_indicator_calculation.py:
```diff
@@ -502,11 +579,177 @@ def equation_degree_of_sector_coupling(
+        Parameters
+        ----------
+        total_generation: float
+            Energy equivalent of total conversion flows
```
generation, not conversion?

Piranias replied:
done

SabineHaas commented on src/multi_vector_simulator/E3_indicator_calculation.py:
```diff
@@ -502,11 +579,177 @@ def equation_degree_of_sector_coupling(
-def equation_onsite_energy_matching():
+def add_onsite_energy_matching(dict_values):
+    """
+    Determines onsite energy matching /self sufficiency and adds KPI to dict_values
```
```suggestion
    Determines onsite energy matching (OEM), i.e. self-sufficiency, and adds KPI to dict_values
```

SabineHaas commented on src/multi_vector_simulator/E3_indicator_calculation.py:
```diff
@@ -502,11 +579,177 @@ def equation_degree_of_sector_coupling(
+        after applying total_renewable_and_non_renewable_energy_origin and
+        total_demand_and_excess_each_sector and
+        add_onsite_energy_fraction
+    Returns
```
```suggestion
    
    Returns
```

SabineHaas commented on src/multi_vector_simulator/E3_indicator_calculation.py:
```diff
@@ -502,11 +579,177 @@ def equation_degree_of_sector_coupling(
+def equation_onsite_energy_matching(
+    total_generation, total_feedin, total_excess, total_demand
+):
+    """Calculates onsite energy matching /self sufficiency.
```
```suggestion
    """
    Calculates onsite energy matching (OEM), i.e. self-sufficiency.
```

SabineHaas reviewed (changes requested):
Thank you @Piranias for adding these KPIs. 

I requested some changes and here are my general comments:
- I would add a section "Notes" before the "Returns" section and move the formulas to "Notes". 
- Please add the test function names (see #todo TEST)
- I don't see how you added weighting in the calculation of the KPIs.
- @smarties would ask you for assertion messages, see E2 tests for examples

Piranias commented on src/multi_vector_simulator/E3_indicator_calculation.py:
```diff
@@ -417,7 +425,75 @@ def equation_renewable_share(total_res, total_non_res):
+        updated dict_values with renewable share of a specific sector
+
+    Tested with
+    - #todo: TEST
```
done

Piranias commented:
> Thank you @Piranias for adding these KPIs.
> 
> I requested some changes and here are my general comments:
> 
>     * I would add a section "Notes" before the "Returns" section and move the formulas to "Notes".

I adjusted the style of the functions to all the other functions in this module. I would not change the styling of single functions now.

> 
>     * Please add the test function names (see #todo TEST)
done
> 
>     * I don't see how you added weighting in the calculation of the KPIs.

I did not add weighting in the functions but I used the "total_flow_of_energy_conversion_equivalent " which should already contain the weighting.  But I'm not 100% sure if I added all the flows correctly, let's wait for @smartie2076 ti check this!

> 
>     * @Smarties would ask you for assertion messages, see E2 tests for examples
done

smartie2076 commented on src/multi_vector_simulator/E3_indicator_calculation.py:
```diff
@@ -502,11 +580,181 @@ def equation_degree_of_sector_coupling(
+            ][CONNECTED_FEEDIN_SINK]
+            + AUTO_SINK
+        ]
+        dso_feedin = dso_sink["total_flow"][0]
```
Please use constant variable instead of `"total_flow"`. Why do you use `dso_sink["total_flow"][0]` and not `dso_sink[TOTAL_FLOW][VALUE]`?

Piranias replied:
I changed this part completely, so this issue is solved now :)

smartie2076 commented on src/multi_vector_simulator/E3_indicator_calculation.py:
```diff
@@ -502,11 +580,181 @@ def equation_degree_of_sector_coupling(
+            + AUTO_SINK
+        ]
+        dso_feedin = dso_sink["total_flow"][0]
+        # sum up all dso flows
+        total_feedin = total_feedin + dso_feedin
```
That only works if we are talking about the same energy carriers - in case of H2+Electricity DSOs you need to apply energy carrier weighting first.

smartie2076 replied:
You could also place this evaluation out of `add_onsite_energy_fration()` into a new function `add_total_feedin`. Name: `TOTAL_FEEDIN + SUFFIX_ELECTRICITY_EQUIVALENT`

Piranias replied:
done

smartie2076 commented on src/multi_vector_simulator/E3_indicator_calculation.py:
```diff
@@ -502,11 +580,181 @@ def equation_degree_of_sector_coupling(
+        {ONSITE_ENERGY_FRACTION: onsite_energy_fraction}
+    )
+    logging.debug(
+        f"Calculated the {ONSITE_ENERGY_FRACTION}: {round(onsite_energy_fraction)}"
```
Round with 2 or 3 decimals

smartie2076 commented on src/multi_vector_simulator/E3_indicator_calculation.py:
```diff
@@ -502,11 +580,181 @@ def equation_degree_of_sector_coupling(
+    - test_equation_onsite_energy_matching()
+    """
+
+    total_generation = (
+        dict_values[KPI][KPI_SCALARS_DICT][TOTAL_RENEWABLE_GENERATION_IN_LES]
+        + dict_values[KPI][KPI_SCALARS_DICT][TOTAL_NON_RENEWABLE_GENERATION_IN_LES]
+    )
```
If you calculate this for each function again and again, wouldnt it be better to add a KPI for this in the function that results in `TOTAL_NON_RENEWABLE_GENERATION_IN_LES `and `TOTAL_NON_RENEWABLE_GENERATION_IN_LES`: `TOTAL_GENERATION_IN_LES`?

Piranias replied:
done

smartie2076 commented on src/multi_vector_simulator/E3_indicator_calculation.py:
```diff
@@ -502,11 +580,181 @@ def equation_degree_of_sector_coupling(
+        {ONSITE_ENERGY_MATCHING: onsite_energy_matching}
+    )
+    logging.debug(
+        f"Calculated the {ONSITE_ENERGY_MATCHING}: {round(onsite_energy_matching)}"
```
Round to 2 or 3 decimals

smartie2076 reviewed (changes requested):
Yay, three more KPI! 

I think a change is necessary for the feedin calculation. 

Also, regarding the tests - I think you only wrote some for the equations, not the `add_xy` functions? 

I see in the tests for the equations that you tested for one example, but not for the extreme values (can autonomy be > 1? What if excessive local generation but also excess for oem?

Piranias commented on src/multi_vector_simulator/E3_indicator_calculation.py:
```diff
@@ -417,7 +425,76 @@ def equation_renewable_share(total_res, total_non_res):
+
+    dict_values[KPI][KPI_SCALARS_DICT].update({DEGREE_OF_AUTONOMY: degree_of_autonomy})
+
+    logging.debug(f"Calculated the {DEGREE_OF_AUTONOMY}: {round(degree_of_autonomy)}")
```
done

Piranias commented on src/multi_vector_simulator/E3_indicator_calculation.py:
```diff
@@ -502,11 +580,181 @@ def equation_degree_of_sector_coupling(
+    total_feedin = 0
+    # sum up all feed into the grid
+    for dso in dict_values[ENERGY_PROVIDERS]:
+        # Get source connected to the specific DSO in question
+        scalars_matrix = dict_values[KPI][KPI_SCALAR_MATRIX]
```
done

smartie2076 commented on src/multi_vector_simulator/E3_indicator_calculation.py:
```diff
@@ -594,29 +643,12 @@ def add_onsite_energy_fraction(dict_values):
-        total_feedin = total_feedin + dso_feedin
-    # save total feedin into KPI_SCALARS_DICT
-    dict_values[KPI][KPI_SCALARS_DICT].update({TOTAL_FEEDIN: total_feedin})
+    add_total_feedin_electricity_equivaluent(dict_values)
```
I would propably call this from `E0`, so that one only has to look at one code section to find the function calls for each of the technical KPI.

Piranias replied:
ok!

smartie2076 reviewed (approved):
Looking good!

Piranias commented:
> Yay, three more KPI!
> 
> I think a change is necessary for the feedin calculation.
> 
> Also, regarding the tests - I think you only wrote some for the equations, not the `add_xy` functions?

I added the missing tests now!
> 
> I see in the tests for the equations that you tested for one example, but not for the extreme values (can autonomy be > 1? What if excessive local generation but also excess for oem?

The degree of autonomy can be > 1. See https://mvs-eland.readthedocs.io/en/latest/MVS_Outputs.html#degree-of-autonomy-da

the OEM cannot be > 1. But this is due to the simulation, I wouldn't know how to test this. For an explanation see: https://mvs-eland.readthedocs.io/en/latest/MVS_Outputs.html#onsite-energy-matching-oem

Bachibouzouk commented:
>     * Check if benchmark tests pass locally (`EXECUTE_TESTS_ON=master pytest`)

@Piranias - did you really run the `EXECUTE_TESTS_ON=master pytest` locally?

I found errors when I did

Piranias commented:
> > ```
> > * Check if benchmark tests pass locally (`EXECUTE_TESTS_ON=master pytest`)
> > ```
> 
> @Piranias - did you really run the `EXECUTE_TESTS_ON=master pytest` locally?
> 
> I found errors when I did

ahh, I did before but forgot to run it again after adding the tests :( sorry. Did you work i out yet, or should I ?

Bachibouzouk commented:
I fixed it in ced0b21bfffabccd95af840336cb7b2031fb7c0c, before we get our CI fixed, make sure you run the tests before you ask for review :)
